pkg/server/plugin/bundlepublisher/awss3/awss3.go

```diff
@@ -0,0 +1,234 @@
+package awss3
+
+import (
+	"bytes"
+	"context"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl"
+	"github.com/spiffe/spire-plugin-sdk/pluginsdk/support/bundleformat"
+	bundlepublisherv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/server/bundlepublisher/v1"
+	"github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/types"
+	configv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/service/common/config/v1"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	pluginName = "aws_s3"
+)
+
+type pluginHooks struct {
+	newS3ClientFunc func(c aws.Config) (simpleStorageService, error)
+}
+
+func BuiltIn() catalog.BuiltIn {
+	return builtin(New())
+}
+
+func New() *Plugin {
+	return newPlugin(newS3Client)
+}
+
+// Config holds the configuration of the plugin.
+type Config struct {
+	AccessKeyID     string `hcl:"access_key_id" json:"access_key_id"`
+	SecretAccessKey string `hcl:"secret_access_key" json:"secret_access_key"`
+	Region          string `hcl:"region" json:"region"`
+	Bucket          string `hcl:"bucket" json:"bucket"`
+	ObjectKey       string `hcl:"object_key" json:"object_key"`
+	Format          string `hcl:"format" json:"format"`
+
+	// bundleFormat is used to store the content of Format, parsed
+	// as bundleformat.Format.
+	bundleFormat bundleformat.Format
+}
+
+// Plugin is the main representation of this bundle publisher plugin.
+type Plugin struct {
+	bundlepublisherv1.UnsafeBundlePublisherServer
+	configv1.UnsafeConfigServer
+
+	config    *Config
+	configMtx sync.RWMutex
+
+	bundle    *types.Bundle
+	bundleMtx sync.RWMutex
+
+	hooks    pluginHooks
+	s3Client simpleStorageService
+	log      hclog.Logger
+}
+
+// SetLogger sets a logger in the plugin.
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure configures the plugin.
+func (p *Plugin) Configure(ctx context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	if req == nil {
```

Comment:
nit: this check isn't practical since a nil request here would imply the gRPC library is broken